pd/src/state/writer.rs

```diff
@@ -259,6 +289,75 @@ impl Writer {
             )
             .execute(&mut dbtx)
             .await?;
+
+            // If the note was previously quarantined, drop it from quarantine
+            query!(
+                "DELETE FROM quarantined_notes WHERE note_commitment = $1",
+                &<[u8; 32]>::from(note_commitment)[..]
+            )
+            .execute(&mut dbtx)
+            .await?;
+        }
+
+        // Calculate the height at which notes quarantined in this block should unbond. If the
+        // unbonding period or the epoch duration change, notes will unbond at the nearest epoch
+        // boundary following this height.
+        let unbonding_epochs = self
+            .private_reader()
+            .chain_params_rx()
+            .borrow()
+            .unbonding_epochs;
+        let epoch_duration = block.epoch.as_ref().unwrap().duration;
+        let unbonding_height = height + (epoch_duration * unbonding_epochs);
+
+        // Add notes and nullifiers from transactions containing undelegations to a quarantine
+        // queue, to be extracted when their unbonding period expires.
+        for QuarantineGroup {
+            validator_identity_key,
+            notes,
+            nullifiers,
+        } in block.quarantine
+        {
+            // Quarantine all notes associated with this quarantine group
+            for (&note_commitment, data) in notes.iter() {
+                // Hold the note data in quarantine
+                query!(
+                    r#"
+                    INSERT INTO quarantined_notes (
+                        note_commitment,
+                        ephemeral_key,
+                        encrypted_note,
+                        transaction_id,
+                        unbonding_height,
+                        validator_identity_key
+                    ) VALUES ($1, $2, $3, $4, $5, $6)"#,
+                    &<[u8; 32]>::from(note_commitment)[..],
+                    &data.ephemeral_key.0[..],
+                    &data.encrypted_note[..],
+                    &data.transaction_id[..],
+                    unbonding_height as i64,
+                    &validator_identity_key.0.to_bytes()[..],
+                )
+                .execute(&mut dbtx)
+                .await?;
+            }
+
+            // Quarantine all nullifiers associated with this quarantine group
+            for &nullifier in nullifiers.iter() {
+                let nullifier_bytes = &<[u8; 32]>::from(nullifier)[..];
+
+                // Keep track of the nullifier associated with the block height
```

Comment:
You might think that we need to also insert the nullifier into the main nullifier set here, but that's taken care of already by this code in `PendingBlock::add_transaction`: https://github.com/penumbra-zone/penumbra/blob/8fc17d4606ad03d5cabb6824606267b9c91966ef/pd/src/pending_block.rs#L177-L181